### PR TITLE
FTL node builder with tests

### PR DESCRIPTION
**Pull request by aaron-prindle**
This is based off of @dlorenc original ftl node builder PR but adds tests.  For now I am reverting to using the tarball format instead of the fromdisk format so I can upload the image rootfs w/ the tests.

### CODE REVIEW

File: ftl/BUILD [hunk 1]
```diff
@@ -43,3 +47,51 @@ py_test(
...
+py_test(
+    name = "node_builder_test",
+    srcs = ["node/builder_test.py"],
+    main = "node/builder_test.py",
+    deps = [
+        ":node_lib",
+        "@mock//:mock",
+    ],
+    data = [
+        ":node/testdata/base_image/config_file",
```
Comment (dlorenc) [inline ftl/BUILD]: you can use glob to make this easier to maintain

File: ftl/node/builder_test.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import tempfile
+
+from containerregistry.client.v2_2 import docker_image
+
+from ftl.common import context
+from ftl.common import cache
+from ftl.node import builder
+
+class NodeTest(unittest.TestCase):
+    def test_create_package_base_uncached(self):
```
Comment (dlorenc) [inline ftl/node/builder_test.py]: nit: you'll need to format most of this to pass travis

File: ftl/node/builder_test.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                cash = cache.MockHybridRegistry('fake.gcr.io/google-appengine', tmp)
+                with b.CreatePackageBase(base_image, cash) as _:
+                    # check that the node_app_layer image was added to the cache
+                    self.assertEqual(1, len(cash.GetMap()))
+                    for k in cash.GetMap():
+                        self.assertEqual(str(k).startswith(
+                            "fake.gcr.io/google-appengine/node-package-lock-cache"
+                            ),
+                        True)
+                cash.ResetCacheMiss()
```
Comment (dlorenc) [inline ftl/node/builder_test.py]: To make this test easier to read, you might want to move most of the boilerplate into a Setup method, then the testcases themselves can be small. You can probably ditch the restCacheMiss API if you switch to several testcases.

File: ftl/BUILD (hunk 1 above)
Comment (aaron-prindle) [reply ftl/BUILD]: done

File: ftl/node/builder_test.py (hunk 3 above)
Comment (aaron-prindle) [reply ftl/node/builder_test.py]: still working on this

**Comment by aaron-prindle** [comment]
What is the buildifier/buildifier.sh for?  It seems to be erroring in travis but it doesn't have any error output.  When I attempt to run the buildifer/buildifer.sh locally I get:
```
$ ./buildifier.sh
./vendor/github.com/docker/distribution/BUILDING.md:6:29: syntax error near intend
```

**Comment by dlorenc** [comment]
Also see the comments Matt left me on https://github.com/GoogleCloudPlatform/runtimes-common/pull/351

Mainly: switch to logging instead of print
add a little bit of extra logging
a few style nits

**Comment by aaron-prindle** [comment]
buildifier.sh is still failing for me after #376

**Comment by aaron-prindle** [comment]
tests are passing now, still to need change to using a setup/parameterized-test-case.  Also should the tests be run in travis?  I don't think they are currently

**Comment by dlorenc** [comment]
> tests are passing now, still to need change to using a setup/parameterized-test-case. Also should the ftl tests be run in travis? I don't think they are currently

Yeah, you should add them to the .travis.yml

**Comment by aaron-prindle** [comment]
added parameterized test and some test classes for node/builder_test.py

**Comment by aaron-prindle** [comment]
added test to .travis.yml.  I believe there is a better way to do this, currently using "bazel run //ftl:node_builder_test" and put node/npm in travis as I don't know how to add node/npm to the linux-sandbox that bazel uses for "bazel test".

File: ftl/common/cache.py [hunk 4]
```diff
@@ -76,14 +76,14 @@ def __init__(self, repo, creds, transport, threads=1, mount=None):
...
+            tag=hashlib.sha256(fingerprint).hexdigest()))
 
     def Get(self, base_image, namespace, checksum):
         entry = self._tag(base_image, namespace, checksum)
-        with docker_image.FromRegistry(
-          entry, self._creds, self._transport) as img:
+        with docker_image.FromRegistry(entry, self._creds,
+                                       self._transport) as img:
             if img.exists():
                 print('Found cached base image: %s.' % entry)
```
Comment (dlorenc) [inline ftl/common/cache.py]: not: logging

File: ftl/common/cache.py [hunk 5]
```diff
@@ -93,6 +93,64 @@ def Get(self, base_image, namespace, checksum):
...
+            namespace=namespace,
+            tag=hashlib.sha256(fingerprint).hexdigest()))
+
+    def Get(self, base_image, namespace, checksum):
+        entry = self._tag(base_image, namespace, checksum)
+        if entry in self._map:
+            return self._map[entry]
+        tarball = os.path.join(self._directory, str(entry))
+        if os.path.isfile(tarball):
+            print('Found cached base image: %s.' % entry)
```
Comment (dlorenc) [inline ftl/common/cache.py]: nit: logging

File: ftl/common/cache.py [hunk 6]
```diff
@@ -93,6 +93,64 @@ def Get(self, base_image, namespace, checksum):
...
+            self._map[entry] = docker_image.FromTarball(tarball)
+            return self._map[entry]
+        print('No cached base image found for entry: %s.' % entry)
```
Comment (dlorenc) [inline ftl/common/cache.py]: logging

File: ftl/node/main.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                transport,
+                threads=_THREADS,
+                mount=[base_name]) as cash:
+            with builder.From(ctx) as bldr:
+                with docker_image.FromRegistry(base_name, base_creds,
+                                               transport) as base_image:
+
+                    # Create (or pull from cache) the base image with the
+                    # package descriptor installation overlaid.
+                    print('Generating dependency layer...')
```
Comment (dlorenc) [inline ftl/node/main.py]: logging

File: ftl/common/cache.py (hunk 6 above)
Comment (aaron-prindle) [reply ftl/common/cache.py]: done

File: ftl/node/main.py (hunk 7 above)
Comment (aaron-prindle) [reply ftl/node/main.py]: done

File: ftl/common/cache.py (hunk 5 above)
Comment (aaron-prindle) [reply ftl/common/cache.py]: done

File: ftl/common/cache.py (hunk 4 above)
Comment (aaron-prindle) [reply ftl/common/cache.py]: done

**Comment by aaron-prindle** [comment]
tests are passing now

**Pull request merged**